promql/query_logger.go

```diff
@@ -0,0 +1,217 @@
+		return false, "[]"
+	}
+
+	var out bytes.Buffer
+	err := json.Indent(&out, []byte(queries), "", "\t")
+
+	if err != nil {
+		level.Warn(logger).Log("msg", "Error parsing json", "err", err)
+		return false, "[]"
+	}
+
+	return true, out.String()
+}
+
+func logUnfinishedQueries(filename string, filesize int, logger log.Logger) {
+	if _, err := os.Stat(filename); err == nil {
+		fd, err := os.Open(filename)
+		if err != nil {
+			level.Error(logger).Log("msg", "Failed to open query log file", "err", err)
+			return
+		}
+
+		brokenJson := make([]byte, filesize)
+		_, err = fd.Read(brokenJson)
+		if err != nil {
+			level.Error(logger).Log("msg", "Failed to read query log file", "err", err)
+			return
+		}
+
+		queriesExist, queries := parseBrokenJson(brokenJson, logger)
+		if !queriesExist {
+			return
+		}
+		level.Info(logger).Log("msg", "These queries didn't finish in prometheus' last run:")
+		fmt.Println(queries)
+	}
+}
+
+func getMMapedFile(filename string, filesize int, logger log.Logger) (error, []byte) {
+
+	fd, err := os.OpenFile(filename, os.O_CREATE|os.O_RDWR|os.O_TRUNC, 0644)
+	if err != nil {
+		level.Error(logger).Log("msg", "Error opening query log file", "file", filename, "err", err)
+		return err, []byte{}
+	}
+
+	if err := fd.Truncate(int64(filesize)); err != nil {
+		level.Error(logger).Log("msg", "Error creating file of specified size", "file", filename, "Attemped size", filesize, "err", err)
+		return err, []byte{}
+	}
+
+	fileAsBytes, err := syscall.Mmap(int(fd.Fd()), 0, filesize, syscall.PROT_READ|syscall.PROT_WRITE, syscall.MAP_SHARED)
+	if err != nil {
+		level.Error(logger).Log("msg", "Error mmapping file", "file", filename, "err", err)
+		return err, []byte{}
+	}
+
+	return err, fileAsBytes
+}
+
+func NewQueryLogger(filename string, filesize int, logger log.Logger) (queryLogger QueryLogger) {
+	defer func() {
+		if r := recover(); r != nil {
+			queryLogger = QueryLogger{}
+		}
+	}()
+
+	logUnfinishedQueries(filename, filesize, logger)
+
+	if filesize < 1+entrySize {
+		level.Error(logger).Log("msg", "Filesize must be greater than entrysize.", "entrysize", entrySize)
+		return QueryLogger{}
+	}
+
+	err, fileAsBytes := getMMapedFile(filename, filesize, logger)
+	if err != nil {
+		return QueryLogger{}
+	}
+
+	copy(fileAsBytes, "[")
+	queryLogger = QueryLogger{
+		mmapedFile:   fileAsBytes,
+		getNextIndex: make(chan int, (1<<16)-1),
+		lastIndex:    make(chan int),
+		logger:       logger,
+		Ok:           true,
+	}
+
+	go queryLogger.generateLastIndex(filesize)
+
+	return queryLogger
+}
+
+type Entry struct {
+	Query     string `json:"query"`
+	Timestamp int64  `json:"timestamp"`
+}
+
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+func newJsonEntry(query string, timestamp int64, logger log.Logger) []byte {
+	// 35 bytes is the size of json entry for empty string. Including a comma for every entry, max query size is entrySize-36
+	entry := Entry{query[:min(len(query), entrySize-36)], timestamp}
```

Comment:
You're measuring characters here, not bytes

Reply:
The smallest entry would be an empty string query, and its size after json.Marshal() turned out to be 35. Since I also want a comma in an entry, and comma is part of UTF-8, (entrySize-36) would be the maximum size of the query, no? Alternatively, I could introduce another variable that repeats this procedure to compute smallest query size in bytes.

Reply:
That's still mixing bytes and characters though.

Reply:
Yes. I spotted my mistake. Using this now: 
```
func trimStringSize(str string, size int) string {
	bytesStr := []byte(str)
	return string(bytesStr[:min(len(str), size)])
}
```